src/ModelIO/MDLAnimatedValueTypes.cs

```diff
@@ -0,0 +1,805 @@
+			if (times == null)
+				throw new ArgumentNullException (nameof (times));
+			if (values.Length != times.Length)
+				throw new InvalidOperationException ($"The length of '{nameof (values)}' array and '{nameof (times)}' array must match.");
+
+			unsafe {
+				fixed (Matrix4d* valuesPtr = values)
+					MDLMemoryHelper.Reset<Matrix4d> ((IntPtr) valuesPtr, times, _ResetWithDouble4x4Array);
+			}
+		}
+
+		public virtual Matrix4 [] GetNMatrix4Values ()
+		{
+			var count = TimeSampleCount;
+			var timesArr = new Matrix4 [(int) count];
+
+			unsafe {
+				fixed (Matrix4* arrptr = timesArr)
+					MDLMemoryHelper.FetchValues<Matrix4> ((IntPtr) arrptr, count, _GetFloat4x4Array);
+			}
+
+			return timesArr;
+		}
+
+		public virtual Matrix4d [] GetNMatrix4dValues ()
+		{
+			var count = TimeSampleCount;
+			var timesArr = new Matrix4d [(int) count];
+
+			unsafe {
+				fixed (Matrix4d* arrptr = timesArr)
+					MDLMemoryHelper.FetchValues<Matrix4d> ((IntPtr) arrptr, count, _GetDouble4x4Array);
+			}
+
+			return timesArr;
+		}
+	}
+
+	public partial class MDLMatrix4x4Array {
+
+		public virtual void SetValues (Matrix4 [] array)
+		{
+			if (array == null)
+				throw new ArgumentNullException (nameof (array));
+
+			unsafe {
+				fixed (Matrix4* arrptr = array)
+					MDLMemoryHelper.SetValues<Matrix4> ((IntPtr) arrptr, array.Length, _SetFloat4x4Array);
+			}
+		}
+
+		public virtual void SetValues (Matrix4d [] array)
+		{
+			if (array == null)
+				throw new ArgumentNullException (nameof (array));
+
+			unsafe {
+				fixed (Matrix4d* arrptr = array)
+					MDLMemoryHelper.SetValues<Matrix4d> ((IntPtr) arrptr, array.Length, _SetDouble4x4Array);
+			}
+		}
+
+		public virtual Matrix4 [] GetNMatrix4Values ()
+		{
+			var count = ElementCount;
+			var timesArr = new Matrix4 [(int) count];
+
+			unsafe {
+				fixed (Matrix4* arrptr = timesArr)
+					MDLMemoryHelper.FetchValues<Matrix4> ((IntPtr) arrptr, count, _GetFloat4x4Array);
+			}
+
+			return timesArr;
+		}
+
+		public virtual Matrix4d [] GetNMatrix4dValues ()
+		{
+			var count = ElementCount;
+			var timesArr = new Matrix4d [(int) count];
+
+			unsafe {
+				fixed (Matrix4d* arrptr = timesArr)
+					MDLMemoryHelper.FetchValues<Matrix4d> ((IntPtr) arrptr, count, _GetDouble4x4Array);
+			}
+
+			return timesArr;
+		}
+	}
+
+	static class MDLMemoryHelper {
+		static IntPtr GetAllocedPtrForArray (Type managedArrayType, int managedArrayCount, out int nativeArrayLength, out IntPtr alignedPtr)
+		{
+			nativeArrayLength = Marshal.SizeOf (managedArrayType) * managedArrayCount;
+			var ptr = Marshal.AllocHGlobal (nativeArrayLength + 16);
+			alignedPtr = new IntPtr (((nint) (ptr + 15) >> 4) << 4);
+			return ptr;
+		}
+
+		internal static void SetValues<T> (IntPtr arrptr, int arrLength, Action<IntPtr, nuint> setFunc)
+		{
+			int nativeArrayLength;
+			IntPtr aligned_ptr;
+
+			var ptr = GetAllocedPtrForArray (typeof (T), arrLength, out nativeArrayLength, out aligned_ptr);
```

Comment:
Something like this would us copy the array only if needed:

```csharp
int nativeArrayLength;
IntPtr aligned_ptr = arrptr;
IntPtr ptr = IntPtr.Zero;

if (aligned_ptr % 16 != 0) {
    ptr = GetAllocatedPtrForArray (typeof (T), arrLength, out nativeArrayLength, out aligned_ptr);
    Runtime.memcpy (aligned_ptr, arrptr, nativeArrayLength);
}

setFunc (aligned_ptr, (nuint) arrLength);

if (ptr != IntPtr.Zero)
    Marshal.FreeHGlobal (ptr);
```